doc/wavetable-communication.md

```diff
@@ -0,0 +1,184 @@
+# Wavetable communication
+
+This explains how RT and non-RT threads communicate when new waves for the
+wavetable need to be generated.
+
+## Wavetable layout
+
+A Tensor3 (3D-Tensor) has the following contents:
+
+```
+  ooooooo 1D-Tensor (frequeny values in Hz)      --
+                                                  |
+  ooooooo 1D-Tensor (semantic values, e.g.        | "wavetable scales"
+                     random seeds or              |
+                     basewave function params)   --
+
+  ooooooo 1D-Tensor (float buffer = 1 wave)      --
+  ^                                               |
+  |                                               | "wavetable data"
+  o...ooo 2D Tensor (Tensor1s for each freq)      | or just
+  ^       with ringbuffer semantics (see below)   | "waves"
+  |       (MW produces, AD note consumes)         |
+  |                                               |
+  o...ooo 3D Tensor (Tensor2s for each semantic) --
+
```

Comment:
Not sure what exactly you mean. The dimensions of the Tensors are fixed. In case we need a Tensor with different dimensions, we create a new Tensor and swap it with the previous one. But I am not sure what you need with flag values.

Reply:
So, there were a number of ideas that were discussed during development and this comment is me trying to figure out which ones came true.

Specifically if you have new wavetables coming in and they're being set to the RT thread you can do pointer swaps, you can do array copies, you can have arrays of pointers and swap out individual ones, etc. So in some of those possible implementation decisions you end up with the RT side 'owning' (i.e. notes on the RT side can use those buffers without locking) buffers that have more memory than the memory that is being actively used.

I'm going to focus on MiddleWare this morning which I hope clarifies these details for me.

Reply:
As this is all done by the RT thread, the code for this is rather in `ADnoteParameters.cpp` in the "set-waves" callback. The function `WaveTable::swapDataAt` is called multiple times there:

```c++
wt->swapDataAt(sem_idx, freq_idx, unusedWave);
```

where `unusedWave` is the raw float array from one wave out of the received blob. So it is rather "individual pointer swaps" on the lowest dimension of the Wavetable.

However, this is just the first part. There are two cases in this callback. In the first case, we just insert one more wave into the ringbuffer, so we do the pointer swap on the active table (`table` or `tableMod`). When loading a completely new table, since we are doing all those swaps bit by bit, we may end up with an incomplete WaveTable. So in this case, this just populates a "next table" (`nextTable` or `nextTableMod`), and only if the current pointer swap finishes building the "next table", the "next table" is being pointer-swapped with the "current table" (`WaveTable::swapWith`, which includes just a couple of pointer swaps on the first dimension).

Reply:
Thanks for the refresher :)